### DISCUSSION TITLE
next.config.js redirect/rewrites need "hasNot" property

### DISCUSSION THREAD

**Discussion by yuvalgershon** [vercel/next.js#25931] [Ideas]
### Describe the feature you'd like to request

My proposition is to add "hasNot" property to the static redirect config.
Exactly the same as "has" but does not redirect if has returned true :)
It would be useful for this to work on all types (Header, Cookie, Query), though I think Header and cookie in particular.
I think this is an obvious and useful tool to have.

My use-case:
I want to statically redirect all unauthenticated users to our login page.
For that I need to detect if the request **does not** container our auth cookie.
Obviously this would be combined with actual auth validation on all requests and redirect.
But static redirect will be faster and cleaner.

### Describe the solution you'd like

adding an "hasnot" field with a list like "has",
or adding "not" property to an "has" object.

### Describe alternatives you've considered

The current recommended method of adding to every page an auth check and redirect is less clean.

ad-hoc solution I tried is two redirects:
1. redirect to same source if has auth cookie.
2. redirect to login always.
where the first one will prevent the second from taking place.
But unfortunately this **doesn't work** due to 1. causing infinite loop.

**Comment by vikasg603** [comment]
Yeah, I also encountered the same issue, but no luck.

**Comment by vitalyliber** [comment]
I'm waiting this feature 👌

**Comment by hacknlove** [comment]
What about using `value: false` instead of `hasNot`

```
module.exports = {
  async redirects() {
    return [
      // if the cookie `authorized` is **not** pressent
      // this redirect will be applied
      {
        source: '/profile/:path*',
        has: [
          {
            type: 'cookie',
            key: 'authorized',
            value: false,
          },
        ],
        permanent: false,
        destination: '/login',
      },
    ]
  },
}
```

Here the PR https://github.com/vercel/next.js/pull/27103

**Reply by yuvalgershon** [reply]
This is a really bad design, firstly a cookie can have a value false, a cookie can have any value.
secondly, it wouldn't solve the case for matching on missing Header or even Query ( Though I am not sure how useful that is ).
I will update the feature request accordingly.

**Reply by hacknlove** [reply]
I think a cookie can have a `"false"` value but not a `false` value, in fact currently `has.value` only accepts strings, so I don't understand your concern.

Also the example is a `cookie` but if you look at the code, you will see it works for headers and query parameters too.

**Reply by yuvalgershon** [reply]
OK I understand sorry :)
I think a separate field like in my proposition, would be much clearer. 
especially to someone not having read the docs.

Your design also doesn't **support matching on not having a specific value**. which was in my original intent.

better support any potential extensions to redirect schema. 
which might desire having support for different value types...
Just a thought.

**Reply by hacknlove** [reply]
Yep. My PR only support "not existing", it does not support "not matching".

I just did the smallest change possible to support the most important missing feature, (on my opinion)

It is a still small change that would support "not matching", to add a field `not: true` inside the `has` objects.

Supporting a `hasNot` field at the `redirect` object seems to me to much effort for something that can be achieved in a more KISS way, anyway I was thinking so far mostly at implementantion level, not at nextjs' developer experience or at future new features.

regarding DX,
I believe `value: null` or `value: false` are semantic enough for the "not existing" feature, which is the one I really need.

regarding future and "not matching" feature, I think declarative config for redirects and rewrites does not need to reach every scenario, but just the more used. In order to support more advanced redirections I would add a function, something like

```
module.exports = {
  async redirects() {
    return [
      {
        source: '/foo',
        customMatch (req) {
           return true; // to redirect, or false to not redirect;
        },
        permanent: false,
        destination: '/bar',
      },
    ]
  },
}
```

This would support anything you want, it's future prof, and it won't add extra complexity to the code, the schema or the documentation.

What do you think?

If it's ok for you I can try to support "customMatch" functions in my PR,

**Reply by yuvalgershon** [reply]
Consider putting custom functions in the "has" condition list instead.
That way you could combine it with simple header/cookie/query matching.

**Reply by hacknlove** [reply]
I added it. 

In a different PR. https://github.com/vercel/next.js/pull/27431

```
      {
        source: '/some-page',
        has: [
          {
            type: 'custom',
            check: (req) => req.cookies.foo === req.query.foo,
          },
        ],
        destination: '/another-page',
      },
```

**Comment by mrlaseptima** [comment]
Someone add this feature, please.
 It's easy to add.😶🙌🙏
